lib/abbreviato/truncated_sax_document.rb

```diff
@@ -0,0 +1,200 @@
+    :tail,
+    :ignored_levels,
+    :truncated
+
+  def initialize(options)
+    @html_coder = HTMLEntities.new
+
+    @max_length = options[:max_length]
+    @tail = options[:tail] || ''
+
+    @truncated_string = ""
+    @closing_tags = []
+    @estimated_length = 0
+    @ignored_levels = 0
+    @truncated = false
+  end
+
+  # This method is called when the parser encounters an open tag
+  def start_element(name, attributes)
+    if max_length_reached? || ignorable_tag?(name)
+      @truncated = true if max_length_reached?
+      return
+    end
+
+    # If already in ignore mode, go in deeper
+    if ignore_mode?
+      enter_ignored_level(name)
+      return
+    end
+
+    string_to_add = opening_tag(name, attributes)
+
+    # Abort if there is not enough space to add the combined opening tag and (potentially) the closing tag
+    length_of_tags = overridden_tag_length(name, string_to_add)
+    if length_of_tags > remaining_length
+      @truncated = true
+      enter_ignored_level(name)
+      return
+    end
+
+    # Save the tag so we can push it on at the end
+    @closing_tags.push name unless single_tag_element?(name)
+
+    append_to_truncated_string(string_to_add, length_of_tags)
+  end
+
+  # This method is called when the parser encounters characters between tags
+  def characters(decoded_string)
+    if max_length_reached? || ignore_mode?
+      @truncated = true
+      return
+    end
+
+    # Use encoded length, so &gt; counts as 4 bytes, not 1 (which is what '>' would give)
+    encoded_string = @html_coder.encode(decoded_string, :named)
+    string_to_append = if encoded_string.bytesize > remaining_length
+      # This is the line which prevents HTML entities getting truncated - treat them as a single char
+      str = @html_coder.encode(truncate_string(decoded_string), :named) || ''
+      str << tail if remaining_length - str.bytesize >= tail.bytesize
+      str
+    else
+      encoded_string
+    end
+    append_to_truncated_string(string_to_append)
+  end
+
+  # This method is called when the parser encounters a comment
+  def comment(string)
+    comment = comment_tag(string)
+    if comment.bytesize <= remaining_length
+      append_to_truncated_string(comment)
+    else
+      @truncated = true
+    end
+  end
+
+  # This method is called when the parser encounters cdata. In practice, this also
+  # gets called for this style of comment inside an element:
+  #
+  #   <style><!--
+  #     /* Font Definitions */
+  #     @font-face
+  #       {font-family:Wingdings;
+  #       panose-1:5 0 0 0 0 0 0 0 0 0;}
+  #   --></style>
+  #
+  def cdata_block(string)
+    if string.bytesize <= remaining_length
+      append_to_truncated_string(string)
+    else
+      @truncated = true
+    end
+  end
+
+  # This method is called when the parser encounters a closing tag
+  def end_element(name)
+    if ignore_mode?
+      exit_ignored_level(name)
+      return
+    end
+    return if max_length_reached? || ignorable_tag?(name)
```

Comment:
It seems like this would be the case when you've already accounted for the closing tag in the `@estimated_length` and it fits exactly in the truncated string. However, since we return early here, you would never get to add that closing tag..